src/main/java/org/jnode/fs/ntfs/datarun/CompressedDataRun.java

```diff
@@ -162,105 +169,201 @@ public int readClusters(long vcn, byte[] dst, int dstOffset, int nrClusters, int
-            // no idea why they even stored it given that it isn't used.
-            // Darwin's version I was referring to doesn't skip this, which seems be a bug.
-            uncompressed.copyFrom(compressed, cpos, 0, len + 1);
-            uncompressed.zero(len + 1, BLOCK_SIZE - 1 - len);
-            cpos++;
-            return len + 3;
+            for (int k = 0; k < len + 1; k++) {
+                uncompressed.put(pos++, compressed.get(cpos++));
+            }
+
+            // If the chunk is uncompressed, the total amount of uncompressed data therein can be computed by adding 1 to this
+            // value (adding 3 bytes to get the total chunk size, then subtracting 2 bytes to account for the chunk
+            // header).
+            return len + 1;
         }
 
-        while (cpos < len + 3 && pos < BLOCK_SIZE) {
+        int rightmostInUncompressed = Math.min(uncompressed.array.length, BLOCK_SIZE);
+
+        // Now this chunk contains compressed data, decompress it.
+        while (cpos < len + 3 && pos < rightmostInUncompressed) {
+            // https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-xca/1fd21d29-f42f-4fc4-b677-9de7cc386be8
+            // If a chunk is compressed, its chunk header is immediately followed by the first byte of a Flag_group nonterminal.
+            //
+            // A flag group consists of a flag byte followed by zero or more data elements.
+            // Each data element is either a single literal byte or a two-byte compressed word.
+            // The individual bits of a flag byte, taken from low-order bits to high-order bits, specify the formats of the subsequent data elements
+            // (such that bit 0 corresponds to the first data element, bit 1 to the second, and so on).
+            // If the bit corresponding to a data element is set, the element is a two-byte compressed word; otherwise, it is a one-byte literal.
             byte ctag = compressed.get(cpos++);
-            for (int i = 0; i < 8 && pos < BLOCK_SIZE; i++) {
+
+            final int bitsInByte = 8;
+            for (int i = 0; i < bitsInByte; i++) {
+
+                // To process compressed buffers,
+                // the size of the compressed chunk that is stored in the chunk header MUST be used
+                // to determine the position of the last valid byte in the chunk.
+                // The size value MUST ignore flag bits that correspond to bytes outside the chunk.
+                if (cpos >= len + 3 || pos >= rightmostInUncompressed) {
+                    break;
+                }
+
                 if ((ctag & 1) != 0) {
                     int j, lmask, dshift;
                     for (j = pos - 1, lmask = 0xFFF, dshift = 12;
-                         j >= 0x10; j >>= 1) {
+                         j >= 0x10; j >>>= 1) {
                         dshift--;
-                        lmask >>= 1;
+                        lmask >>>= 1;
```

Comment:
Interesting this change to use >>> which makes sense.  It makes me wonder if there is any other code in thius module thta might be using >> instead of >>>.  Have you checked?